ostd/src/mm/kspace/kva.rs

```diff
@@ -0,0 +1,456 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Kernel virtual memory allocation
+
+use alloc::{
+    collections::{vec_deque::VecDeque, BTreeMap},
+    vec::Vec,
+};
+use core::ops::{DerefMut, Range};
+
+use align_ext::AlignExt;
+pub(crate) use lazy_static::lazy_static;
+
+use super::{KERNEL_PAGE_TABLE, TRACKED_MAPPED_PAGES_RANGE, VMALLOC_VADDR_RANGE};
+use crate::{
+    arch::mm::tlb_flush_addr_range,
+    cpu::{num_cpus, CpuSet, PinCurrentCpu},
+    cpu_local,
+    mm::{
+        page::{
+            meta::{PageMeta, PageUsage},
+            DynPage, Page,
+        },
+        page_prop::{CachePolicy, PageFlags, PageProperty, PrivilegedPageFlags},
+        page_table::PageTableItem,
+        FallibleVmRead, FallibleVmWrite, Vaddr, VmIo, VmReader, VmWriter, PAGE_SIZE,
+    },
+    sync::SpinLock,
+    task::disable_preempt,
+    Error, Result,
+};
+
+pub struct KvaFreeNode {
+    block: Range<Vaddr>,
+}
+
+impl KvaFreeNode {
+    pub(crate) const fn new(range: Range<Vaddr>) -> Self {
+        Self { block: range }
+    }
+}
+
+pub struct VirtAddrAllocator {
+    freelist: BTreeMap<Vaddr, KvaFreeNode>,
+}
+
+impl VirtAddrAllocator {
+    fn new(range: Range<Vaddr>) -> Self {
+        let mut freelist: BTreeMap<Vaddr, KvaFreeNode> = BTreeMap::new();
+        freelist.insert(range.start, KvaFreeNode::new(range));
+        Self { freelist }
+    }
+
+    /// Allocate a kernel virtual area.
+    ///
+    /// This is currently implemented with a simple FIRST-FIT algorithm.
+    fn alloc(&mut self, size: usize) -> Result<Range<Vaddr>> {
+        let mut allocate_range = None;
+        let mut to_remove = None;
+
+        for (key, value) in self.freelist.iter() {
+            if value.block.end - value.block.start >= size {
+                allocate_range = Some((value.block.end - size)..value.block.end);
+                to_remove = Some(*key);
+                break;
+            }
+        }
+
+        if let Some(key) = to_remove {
+            if let Some(freenode) = self.freelist.get_mut(&key) {
+                if freenode.block.end - size == freenode.block.start {
+                    self.freelist.remove(&key);
+                } else {
+                    freenode.block.end -= size;
+                }
+            }
+        }
+
+        if let Some(range) = allocate_range {
+            Ok(range)
+        } else {
+            Err(Error::KvaAllocError)
+        }
+    }
+
+    /// Free a kernel virtual area.
+    fn free(&mut self, range: Range<Vaddr>) {
```

Comment:
A better option for this singleton API design is to allow no-locking patterns. It is better to make every `&mut` self APIs `&self` and let the `VirtAddrAllocator` manages it's lock scheme internally. Then in the future we can easily support per-cpu locks without modifying the user code.